crypto/thread/thread_external.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright 1995-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/e_os2.h>
+
+#if defined(OPENSSL_THREADS)
+# include <openssl/crypto.h>
+
+# include "e_os.h"
+
+#  if defined(_WIN32)
+#   include <windows.h>
+#  endif
+
+# include <openssl/async.h>
+
+# include "thread.h"
+# include "thread_external.h"
+# include "internal/list.h"
+
+/* The following is required to keep ABI compat with CRYPTO_THREAD_ROUTINE
+ * as defined in openssl/crypto.h */
+# if !defined(CALLBACK)
+#  if defined(_WIN32) || defined(_STDCALL_SUPPORTED)
+#   define CALLBACK __stdcall
+#  else
+#   define CALLBACK
+#  endif
+# elif !defined(_WIN32) && !defined(_STDCALL_SUPPORTED)
+#  undef CALLBACK
+# endif
+
+volatile int CRYPTO_THREAD_EXTERN_enabled = 0;
+
+# ifdef OPENSSL_NO_EXTERN_THREAD
+
+int CRYPTO_THREAD_EXTERN_enable(CRYPTO_SIGNAL** props)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_EXTERN_disable()
+{
+    return 1;
+}
+
+# else /* ! OPENSSL_NO_EXTERN_THREAD */
+
+static struct list    CRYPTO_THREAD_EXTERN_task_queue;
+static struct list    CRYPTO_THREAD_EXTERN_task_done;
+static CRYPTO_MUTEX   CRYPTO_THREAD_EXTERN_task_lock;
+static CRYPTO_CONDVAR CRYPTO_THREAD_EXTERN_task_cond_create;
+static CRYPTO_CONDVAR CRYPTO_THREAD_EXTERN_task_cond_finish;
```

Comment:
All of these global variables should be stored in the OPENSSL_CTX instead, and then we should pass the OPENSSL_CTX to the various thread functions that need it.